Species-tree parsing

mmatschiner opened the pull request:
Parsing of species trees in Newick or Nexus format. If in the latter and written by StarBEAST, population-size estimates per branch are used in the definition of population configurations.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+
+        # Parse the bifurcating part of the tree.
+        err = 'It appears that the tree string does not include branch lengths!'
+        assert ":" in working_newick_string, err
```
We should raise a ValueError here rather than an assert, I think. I think asserts are for internal sanity checks ("this assumption I'm making should definitely be true"), not input validation. So, this would be changed to:
```python
if ":" not in working_newick_string:
    raise ValueError("Species tree strings must include branch lenghts")
```

jeromekelleher replied:
There's a bunch of other places with the same pattern, I won't comment on them.

mmatschiner replied:
Thanks, I didn't know that. Should now be fixed in my fork.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+                nstsp = newick_string_tail_start_pos
+                working_newick_string = working_newick_string[1:nstsp+1]
+
+        # Parse the bifurcating part of the tree.
```
does it have to be strictly binary? What happens if someone passes in a tree with more than 2 children at a node?

mmatschiner replied:
It does. The parser throws an error when a node is not binary. Unless you think it should be changed,  I'ld prefer to keep this restriction to binary trees, because in practice, species trees inferred with all or at least most methods are binary, and because I'm a bit worried that changes to the parser might make it less robust.

petrelharp replied:
I vote for not restricting to binary trees. What sort of non-robustness are you worried about?

jeromekelleher replied:
Let's keep it to binary trees for the initial version then, and see about generalising later. The restriction is fine I think, once we raise a good error when we encounter a non-binary tree.

mmatschiner replied:
By using the newick package for parsing now, binary trees are allowed. They are randomly resolved by injecting zero-length branches.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+        max_total_edge_length = max(total_edge_lengths)
+        if (max_total_edge_length - min(total_edge_lengths) >
+                0.01 * max_total_edge_length):
+            raise NotImplementedError(
```
This is a ValueError --- NotImplementedError is used when a subclass hasn't implemented an abstract method from a superclass

mmatschiner replied:
Fixed.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+        else:
+            return True
+
+    def set_length(self, length):
```
I don't think we need all the Java-style getters and setters here, we can just access and set the ``length`` attribute directly (particularly since this is an internal API, and won't be exposed to the user)

mmatschiner replied:
Ok, thanks. All these are no longer included.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+    def get_node_ids(self):
+        return self.node_ids
+
+    def get_is_extant(self):
```
I think we can implement this as a property:
```python
@property()
def is_extant(self):
    return self.node_ids[1][0:12] != 'internalNode'
```

mmatschiner replied:
This also is no longer included.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+    if branch_length_units == "gen":
+        generations_per_branch_length_unit = 1
+    elif branch_length_units == "myr":
+        generations_per_branch_length_unit = 1000000/generation_time
```
10**6 is more readable here

mmatschiner replied:
Fixed.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+
+    # Read the input file.
+    tree_string = None
+    starbeast_format = False
```
I don't think we should try to detect the format --- it's much simpler and more robust to make user do it (they'll know the format, presumably). 

I think we should split this into two functions: ``parse_species_tree`` and ``parse_starbeast``. That would simplify the parameterisation (we don't have to explain that one parameter is expected if it's in one format but not if it's in the other, etc), and make the code more readable.

mmatschiner replied:
I now split the code into these two functions.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+    inlines = infile.readlines()
+    assert len(inlines) > 0, 'File {} is empty.'.format(filename)
+    if inlines[0][0:6].lower() == '#nexus':
+        # Assume the input is in nexus format. Maximally one tree string is read.
```
Let's split this out into a function, ``parse_nexus``

mmatschiner replied:
The new functions `parse_species_tree` and `parse_starbeast` now allow input that is either a one-line Newick string or a Nexus file written by StarBEAST. Nexus files that are not in StarBEAST format are not allowed. If we stick to this, the change that @jeromekelleher suggests here is no longer relevant.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+
+
+def parse_species_tree(
+        filename=None,
```
I think it would be simpler for now to pass a string as the argument rather than insisting on a file path. It makes things pretty tedious for testing if we have to write a bunch of tempfiles. It's easy enough for the user to write
```python
with open("treefile", "r") as f:
     parse_species_tree(f.read())
```

We can add an additional ``filename`` argument also later if you like, but for the initial version let's get it working with a string parameter.

mmatschiner replied:
I agree, this is now changed.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+    internal_pop_sizes = []
+    divergence_times = []
+    for edge in species_tree.get_edges():
+        if edge.get_is_extant() is False:
```
It's not a good idea to compare booleans with object identity. 
```python
if edge.is_extant:
```
is more idiomatic.

mmatschiner replied:
Thanks, this is now changed.

jeromekelleher commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,834 @@
+                proportion=1.0))
+
+    # Return a tuple of population_configurations and demographic_events.
+    return (population_configurations, demographic_events)
```
The brackets aren't needed here.

mmatschiner replied:
Fixed.

jeromekelleher commented on msprime/__init__.py:
```diff
@@ -51,3 +51,4 @@
 from msprime.exceptions import *
 from msprime.mutations import *
 from msprime.likelihood import *
+from msprime.species_tree import *
```
Not sure we want to import everything into the msprime namespace here, as there's going to be quite a bit of private stuff in here we don't want the user to see. I think we should import ``parse_species_tree`` and parse_startbeast`` here.

mmatschiner replied:
Makes sense!

mmatschiner replied:
Now changed.

jeromekelleher commented on tests/data/species_trees/primates.tre:
```diff
@@ -0,0 +1 @@
+(((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)
```
Not really worth storing these as files; we could just keep them as module level variables in ``test_species_tree.py``, like
```python
primates_tree = "(((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)"
```

mmatschiner replied:
Done, except for two large input trees (one in starbeast format, one with over 100 species).

jeromekelleher commented:
I've had a bit of a play with this @mmatschiner, and I think our lives would be a lot simpler if we used the [newick](https://pypi.org/project/newick/) package to do the parsing. I inserted this into the ``parse_extended_newick_string``:
```python
        print("self.newick_string = ", self.newick_string)
        root = newick.loads(self.newick_string)[0]
        for node in root.walk():
            print(node.length, node.name)
```
and we get 
```
self.newick_string =  ((((((spc12[&dmv={0.11865530745434695}]:2.1888253779862743,spc11[&dmv={0.3619158652399699}]:2.1888253779862743)[&dmv={0.8111402760207012}]:0.5650137097072183,((spc08[&dmv={0.15615846438202152}]:1.1573795518978671,spc06[&dmv={0.449247825255664}]:1.1573795518978671)[&dmv={0.22731541596814978}]:0.7339367424408518,(spc07[&dmv={0.7068472628949976}]:0.06103616637684422,spc05[&dmv={0.6657390159580989}]:0.06103616637684422)[&dmv={0.08988914060435668}]:1.8302801279618748)[&dmv={0.25563233175600886}]:0.8625227933547737)[&dmv={0.3262773143035577}]:4.492254220433834,(spc04[&dmv={0.2580836901510064}]:5.7741246038893665,spc03[&dmv={0.21825140538284654}]:5.7741246038893665)[&dmv={0.23175238045171034}]:1.47196870423796)[&dmv={0.16637277604025583}]:1.3138908669207447,spc02[&dmv={0.3822654051014509}]:8.55998417504807)[&dmv={0.32647418064987677}]:9.417574569831729,(spc09[&dmv={0.1799564092258708}]:16.063369096903575,spc01[&dmv={0.02494323374786103}]:16.063369096903575)[&dmv={0.6254159122303401}]:1.914189647976226)[&dmv={1.1800837491137275}]:3.2008553127180406,spc10[&dmv={0.019049791450860193}]:21.17841405759784)
0.0 None
3.2008553127180406 [&dmv={1.1800837491137275}]
9.417574569831729 [&dmv={0.32647418064987677}]
1.3138908669207447 [&dmv={0.16637277604025583}]
4.492254220433834 [&dmv={0.3262773143035577}]
0.5650137097072183 [&dmv={0.8111402760207012}]
2.1888253779862743 spc12[&dmv={0.11865530745434695}]
2.1888253779862743 spc11[&dmv={0.3619158652399699}]
0.8625227933547737 [&dmv={0.25563233175600886}]
0.7339367424408518 [&dmv={0.22731541596814978}]
1.1573795518978671 spc08[&dmv={0.15615846438202152}]
1.1573795518978671 spc06[&dmv={0.449247825255664}]
1.8302801279618748 [&dmv={0.08988914060435668}]
0.06103616637684422 spc07[&dmv={0.7068472628949976}]
0.06103616637684422 spc05[&dmv={0.6657390159580989}]
1.47196870423796 [&dmv={0.23175238045171034}]
5.7741246038893665 spc04[&dmv={0.2580836901510064}]
5.7741246038893665 spc03[&dmv={0.21825140538284654}]
8.55998417504807 spc02[&dmv={0.3822654051014509}]
1.914189647976226 [&dmv={0.6254159122303401}]
16.063369096903575 spc09[&dmv={0.1799564092258708}]
16.063369096903575 spc01[&dmv={0.02494323374786103}]
21.17841405759784 spc10[&dmv={0.019049791450860193}]
```

So it seems like we can just parse the extended newick annotations out of the ``node.name`` attribute? This seems a **lot** easier to me than parsing newick...

mmatschiner commented:
That's right. I'll rewrite species_tree.py to use the newick package.

mmatschiner commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,489 @@
+    root.remove_redundant_nodes()
+
+    # Resolve polytomies by injecting zero-length branches.
+    root.resolve_polytomies()
```
It seems `resolve_polytomies()` isn't exactly doing what I thought it would. Not sure if the ancestor property is not getting updated as it should:
```
import newick

def print_node_heights(tree):
    root = tree
    # Set node depths (distances from root)
    max_depth = 0
    for node in root.walk():
        depth = 0
        moving_node = node
        while moving_node.ancestor:
            depth += moving_node.length
            moving_node = moving_node.ancestor
        node.depth = depth
        if depth > max_depth:
            max_depth = depth
    # Set node heights (distances from present).
    for node in root.walk():
        node.height = max_depth - node.depth
        print(node.height)

root = newick.loads("((human:1,chimpanzee:1,gorilla:1):2,orangutan:3)")[0]
print_node_heights(root)
print()
root.resolve_polytomies()
print_node_heights(root)
```
Anyway, the best way to avoid this issue may be not to use resolve_polytomies, so I'll follow your advice and remove the dependence on bifurcating trees (at least for the newick parsing, not for the starbeast parsing though).

mmatschiner replied:
The output of the above code is:
```
3.0
1.0
0.0
0.0
0.0
0.0

```(before calling resolve_polytomies)
and 
```
3.0
1.0
0.0
3.0
2.0
2.0
0.0

```
(after calling it). So somehow nodes end up having a height of  2 (which is not in the input tree) and fewer have a height of 0 (even though four nodes are extant).

mmatschiner replied:
Ok, resolve_polytomies() is no longer used, and neither is remove_redundant_nodes(). Instead the code now works with trees that have nodes with any number of descendants.

mmatschiner commented on msprime/species_tree.py:
```diff
@@ -0,0 +1,489 @@
+    return generations_per_branch_length_unit
+
+
+def get_species_tree_string(species_tree_line=None):
```
I now removed the checks for equal numbers of opening and closing parentheses and for having at least one parenthesis open throughout the string. The error messages that the newick package throws could be better, but I agree, let's leave this responsability to that package.

mmatschiner commented:
Tests are going to come next.

jeromekelleher commented:
Looking really nice now @mmatschiner, thanks. I'm looking forward to the tests (as I say, I can help out a bit here, just let me know).

jeromekelleher commented:
Tests looks good @mmatschiner, probably needs a bit more on parsing bad data though.

I can take it from here if you like?

mmatschiner commented:
Feel free to! I'm new to testing, but I did what seemed to make sense for me. I'll leave this to you now but please let me know if I should do anything else.

jeromekelleher commented:
OK, cool, I'll update things here soon and ping you for a review.

Just so you know, I'll need to write to your fork's master branch when I'm [rebasing](https://stdpopsim.readthedocs.io/en/latest/development.html#rebasing), so it will get out of sync with your local version. This is one of the reasons that using [topic branches](https://stdpopsim.readthedocs.io/en/latest/development.html#github-workflow) is a good idea. You basically then never use the master branch of your fork, so don't need to worry about keeping it synced.

Anyway, doesn't make any difference here, just wanted to give you a heads up in case you encountered weirdness later.

jeromekelleher commented:
I've spent a good bit of time with this @mmatschiner, and updated to simplify the algorithm a bit and to update the tests. Hopefully this is an improvement. I think that the tests for parse_species_tree are pretty good now (although we do need to make sure we're round tripping correctly with generation time units). There's still a lot of work to be done in parse_startbeast though. There's a **lot** of code here, and a lot of places for things to go wrong, and we need to test and verify all those places. Hopefully what I've done here will serve as a useful platform for updating the code with some more tests.

Here's what we need to do I think:

1. Test round tripping with different branch_length_units and generation_time values
2. Lots of tests for malformed nexus intput for starbeast. The newick side of things is pretty well covered, as we're using the same code now for both, but we need to verify what happens when we have malformed translation blocks and all sorts of stuff. It's messy.
3. Check for malformed annotations in starbeast, plus unparseable dmv values. Also those where dmv is present for only some nodes (so the initial regexp doesn't catch the error)
4. Test that we're doing the back translating species names correctly when doing translations (although, we're not actually using the names now...)
5. Probably add ``metadata={"species_name": node.name}`` into the PopulationConfiguration object so that we can track that we're capturing the structure correctly (and make the information available to the user)
6. Document the resulting simulation more clearly. Basically we need to paraphrase the translation algorithm so that users will know what each population means at each point in time. I.e, we always merge into the left-most population, which then becomes the ancestral pop.

That's about it I think. Apologies is this seems like a lot more work, but parsing text well is hard, and it needs a lot of testing. You should be able to add more commits to this branch if you run ``git pull`` locally, and then push new commits as before.

mmatschiner commented:
Hi Jerome, thanks a lot for all the efforts. I haven't had a chance yet to go through everything, sorry for that. Regarding StarBEAST tree parsing; I see two things to consider:

1. We could assume that StarBEAST trees are really just written by the program StarBEAST and not manually by the user or by other tools. If we only expect trees written by StarBEAST, we know how they will be written (https://github.com/genomescale/starbeast2/blob/master/src/starbeast2/SpeciesTreeLogger.java) and we would not need to worry too much about malformed input.

2. If you think the StarBEAST part is too difficult to test and maintain properly I wouldn't insist on keeping it in msprime. Would it make more sense to distribute the function with StarBEAST instead so the responsability is not with the msprime developers?

I'll work on addressing your comments over the next few days.

jeromekelleher commented:
Your point about starbeast input being machine generated is a good one @mmatschiner, so I guess we don't need to be *too* worried about raising error messages and just using lots of ``asserts`` to sanity check is a good approach. 

One thing we could do then is to be quite strict about **only** accepting stuff that has come from starbeast (for example, does it always use a translation block?), which could simplify the code a bit too.

What do you think?

mmatschiner commented:
Good, I agree. I'll contact the StarBEAST developers about our plans to check which aspects of the format they consider stable.

jeromekelleher commented:
Hey @mmatschiner, we'd like to get this merged pretty soon. Perhaps we could get the small things finished up, and then open some issues for things like improving starbeast test coverage?

mmatschiner commented:
Sorry for the delay. A number of other projects unexpectedly need my attention. But I'll prioritize this now. Your suggestion sounds fine!

mmatschiner commented:
Took me a bit to figure out how to modify tree object, but https://tskit.readthedocs.io/en/latest/tutorial.html#working-with-tables helped.

mmatschiner commented:
Addressing point 6 next...

jeromekelleher commented:
Great, please ping me when you'd like input @mmatschiner.

mmatschiner commented:
Following https://github.com/tskit-dev/msprime/pull/893#issuecomment-597548838 I added asserts for parse_starbeast. Please @jeromekelleher let me know if you think this is not going in the right direction. Otherwise I'll continue with
1. extending TestStarbeastRoundTrip
2. adding more asserts to the translation block parsing.

Additionally, I wonder if we should add something like this:
```
        for y in range(len(root.get_leaves())):
            if y != source:
                demographic_events.append(
                    msprime.MigrationRateChange(
                        time=divergence_times[x]*generations_per_branch_length_unit,
                        rate=0,
                        matrix_index=(source, y)))
                demographic_events.append(
                    msprime.MigrationRateChange(
                        time=divergence_times[x]*generations_per_branch_length_unit,
                        rate=0,
                        matrix_index=(y, source)))

```
with every MassMigration. It would facilitate the use of migration rate matrices by stopping migration into lineages after (before in forward time) the MassMigration. On the other hand, it would implicitly assign the migration rates of the left-most childs to ancestral species which may also not be wanted.

jeromekelleher commented:
The changes look good, thanks @mmatschiner, I think you're going in the right direction. Let's not worry about migration rates for now, we can do this in a follow up issue I think.

mmatschiner commented:
I think points 2-4 and 6 of https://github.com/tskit-dev/msprime/pull/893#issuecomment-595210458 are now also addressed. Let me know if anything more is missing, @jeromekelleher.

jeromekelleher commented:
Great, thanks @mmatschiner. I'll finish things up from here and open issues for any stuff that I think needs following up on.
